### PR TITLE
Disallow ALTER ROLE in PG for Babelfish-created roles

### PR DESCRIPTION
**Pull request by Suthapalli-Ramya-satya-vasavi-srija**
Signed-off-by: vasavi suthapalli <svasusri@amazon.com>

### Description

Previously, for a PG user to break a Babelfish instance by using ALTER ROLE to block logins, change role membership, change passwords, and other changes that would have a dramatic effect on Babelfish functionality. Now, ALTER ROLE been disabled for Babelfish-created PG roles.
 
### Issues Resolved
Task: Babel-2976

### Testing scenarios covered
- Use case based - Tested and added in the description
- Boundary conditions - Tested whether the error is being thrown when not enable_drop_babelfish_role and successfully dropping/altering when enable_drop_babelfish_role is enabled.
- Arbitrary inputs - Tested lower/upper/mixed case input for bbf created roles/users, db
- Negative test cases - Tested whether able to throw error why we try to alter/drop bbf roles/users, db
- Minor version upgrade tests - N/A
- Major version upgrade tests - N/A
- Performance tests - N/A
- Tooling impact - N/A
- Client tests - N/A
### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by xhfanhe** [comment]
Please add test to upgrade schdules

**Comment by xhfanhe** [comment]
This test doesn't need to be an upgrade test.

**Comment by shalinilohia50** [comment]
Please add test template to the PR description.

File: contrib/babelfishpg_tds/src/backend/tds/tdsutils.c [hunk 1]
```diff
@@ -1146,3 +1155,72 @@ is_babelfish_ownership_enabled(ArrayType *array)
...
+        return true;
+
+    check_babelfish_alterrole_restictions(name);
+
+    /*
+     * Reaching here does not mean that this user has permission to modify the role.
+     * Those permissions checks are done through normal handling.
+     */
+    pfree(name);
+    name = NULL;
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tds/src/backend/tds/tdsutils.c]: We can remove this line. No need to assign NULL.

File: contrib/babelfishpg_tds/src/backend/tds/tdsutils.c [hunk 2]
```diff
@@ -1146,3 +1155,72 @@ is_babelfish_ownership_enabled(ArrayType *array)
...
+
+    /* If the role does not exist, just let the normal Postgres checks happen. */
+    if (name == NULL)
+        return true;
+
+    check_babelfish_alterrole_restictions(name);
+
+    /* We don't need "name" anymore */
+    pfree(name);
+    name = NULL;
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tds/src/backend/tds/tdsutils.c]: We can remove this line. No need to assign NULL.

File: contrib/babelfishpg_tds/src/backend/tds/tdsutils.c [hunk 3]
```diff
@@ -696,6 +699,12 @@ void tdsutils_ProcessUtility(PlannedStmt *pstmt,
 		case T_RenameStmt:
 			handle_result = handle_rename((RenameStmt *)parsetree);
 			break;
+		case T_AlterRoleStmt:
+			handle_result = handle_alter_role((AlterRoleStmt*)parsetree);
+			break;
+		case T_AlterRoleSetStmt:
+			handle_result = handle_alter_role_set((AlterRoleSetStmt*)parsetree);
+			break;
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tds/src/backend/tds/tdsutils.c:707]: Is this change needed now? Won't it be handled by OAT_POST_ALTER case in object_access_hook?

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

File: test/JDBC/input/ownership-restrictions-from-pg.mix [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+GO
+
+DROP ROLE r3;
+GO
+
+SET enable_drop_babelfish_role = false;
+GO
+
+
+CREATE ROLE r4;
```
Comment (shalinilohia50) [inline test/JDBC/input/ownership-restrictions-from-pg.mix]: I think adding file names as prefixes to the names would be better and helpful in debugging the issues later(if any).

Comment (Suthapalli-Ramya-satya-vasavi-srija) [reply test/JDBC/input/ownership-restrictions-from-pg.mix]: Fixed in next revision by adding file name as prefixes for role/user/login names.

**Review by xhfanhe — APPROVED** [review]

**Pull request merged**